rbac/management/group/view.py

```diff
@@ -340,3 +348,126 @@ def principals(self, request, uuid=None):
+
+        @apiParam (Path) {String} id Group unique identifier.
+
+        @apiSuccess {String} uuid Group unique identifier
+        @apiSuccess {String} name Group name
+        @apiSuccess {Array} roles Array of roles
+        @apiSuccessExample {json} Success-Response:
+            HTTP/1.1 200 OK
+            {
+                "roles": [
+                    {
+                        "name": "RoleA",
+                        "uuid": "4df211e0-2d88-49a4-8802-728630224d15",
+                        "description": "RoleA Description"
+                    }
+                ]
+            }
+        """
+        """
+        @api {post} /api/v1/groups/:uuid/roles/   Add roles to a group
+        @apiName addRoles
+        @apiGroup Group
+        @apiVersion 1.0.0
+        @apiDescription Add roles to a group
+        @apiHeader {String} token User authorization token
+        @apiParam (Path) {String} id Group unique identifier
+        @apiParam (Request Body) {Array} roles Array of role UUIDs
+        @apiParamExample {json} Request Body:
+            {
+                "roles": [
+                    "4df211e0-2d88-49a4-8802-728630224d15"
+                ]
+            }
+        @apiSuccess {String} uuid Group unique identifier
+        @apiSuccess {String} name Group name
+        @apiSuccess {Array} roles Array of roles
+        @apiSuccessExample {json} Success-Response:
+            HTTP/1.1 200 OK
+            {
+                "roles": [
+                    {
+                        "name": "RoleA",
+                        "uuid": "4df211e0-2d88-49a4-8802-728630224d15",
+                        "description": "RoleA Description"
+                    }
+                ]
+            }
+        """
+        """
+        @api {delete} /api/v1/groups/:uuid/roles/   Remove roles from group
+        @apiName removeRoles
+        @apiGroup Group
+        @apiVersion 1.0.0
+        @apiDescription Remove roles from a group
+
+        @apiHeader {String} token User authorization token
+
+        @apiParam (Path) {String} id Group unique identifier
+
+        @apiParam (Query) {String} roles List of comma separated role UUIDs
+
+        @apiSuccessExample {json} Success-Response:
+            HTTP/1.1 204 NO CONTENT
+        """
+        roles = request.data.pop(ROLES_KEY, [])
+        group = self.get_object()
+        if request.method == 'POST':
+            self.add_roles(group, roles)
+            response_data = GroupRoleSerializer(group)
+        elif request.method == 'GET':
+            response_data = GroupRoleSerializer(group)
+        else:
+            if ROLES_KEY not in request.query_params:
+                key = 'detail'
+                message = 'Query parameter {} is required.'.format(ROLES_KEY)
+                raise serializers.ValidationError({key: _(message)})
+            role_ids = request.query_params.get(ROLES_KEY, '')
+
+            self.remove_roles(group, role_ids.split(','))
```

Comment:
@chambridge I did test this locally (I'll add some tests explicitly for this per your other comment 👍) and that's what the `ValidationError` catches, actually. Both in `add_roles` and `remove_roles`. Otherwise, the following is raised if you try to add or remove a non-UUID:
```
django.core.exceptions.ValidationError: ["'asdf' is not a valid UUID."]
```
Do you think that's explicit enough, or should we have another check to throw away non-UUIDs altogether?

Reply:
I see, serializers have a `ValidationError` as well so that didn't jump out a the above exception type.

Up to you; but instead of just logging: 
```
logger.info('No role with id %s to delete from group.', role_id)
```
You could test the id value is a uuid and if not respond with a 400 (i.e. `serializers.ValidationError(error)` ); so the user knows they provided bad input (additionally there is a potential savings of not querying the database).

Reply:
@chambridge I was wondering about that; the case where a subset of the IDs being passed in are either not UUIDs, or are not found in the DB, but are valid (either for a POST or DELETE). So say you want to add the following:

```
{
  "roles": [
    "57e60f90-8c0c-4bd1-87a0-2143759aae1c",
    "395e64e2-a981-496c-bb00-833dc9b14cf9",
    "abc123"
  ]
}
```
Where: 
`57e60f90-8c0c-4bd1-87a0-2143759aae1c` is valid and exists in the DB
`395e64e2-a981-496c-bb00-833dc9b14cf9` is valid but does **not** exist in the DB
`abc123` is invalid

Is the expectation that we fail the entire request with a 400? Ignore the invalid ids and return a 200? Have a multi-status 207 with additional info about what failed?

I was going to ask you if there were any patterns for similar already in place. It looked like at least for principal deletion from a group, we're currently just logging the exception and throwing away the username, continuing with the rest of the list and returning a 200. Create is a bit different, as it finds or creates the principal. That was really the primary reason; to keep the interface consistent.

Reply:
My initial reaction is just do a 400, if any of the input is bad and fail the whole request. 

If the input is of the right format, then typically the code currently just logs the failure with a 200 (but this was mostly special case, because of the need to interact with the user service). I can see how a 207 could be an improvement, but I feel like user experience would be a good driver here as to how much effort to expend.

When deleting, I'm in favor of ignoring missing entries; since the intent is to remove them.
When adding it is more complex and I think you either ignore or 207 (with ignore being the crawl step).